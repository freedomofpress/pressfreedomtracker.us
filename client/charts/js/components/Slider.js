import React, { useState, useRef, useEffect } from 'react'
import * as d3 from 'd3'
import { clamp, first, last } from 'lodash'

export function Slider({ elements, xScale, y, setSliderSelection, sliderSelection, idContainer }) {
  const onSliderReleaseRef = useRef(null)
  const onMouseMoveRef = useRef(null)
  const [mousePosition, setMousePosition] = useState({ x: null, y: null })

  useEffect(() => {
    setSliderSelection(first(elements))
  }, [first(elements)])

  function stopMovingSlider(event) {
    window.removeEventListener('mouseup', onSliderReleaseRef.current)
    onSliderReleaseRef.current = null

    window.removeEventListener('mousemove', onMouseMoveRef.current)
    onMouseMoveRef.current = null

    setMousePosition({ x: null, y: null })
  }

  function followMouse(event) {
    const svg = document.getElementById(idContainer)
    if (svg === null) return

    const dim = svg.getBoundingClientRect()
    const mouse = { x: event.clientX - dim.left, y: event.clientY - dim.top }
    setMousePosition(mouse)

    const distancesToMouse = elements.map((d) => ({
      x: d,
      distance: Math.abs(xScale(d) - mouse.x),
    }))
    const selectedValue = distancesToMouse.find(
      (d) => d.distance === d3.min(distancesToMouse.map((e) => e.distance))
    ).x
    setSliderSelection(selectedValue)
  }

  return (
    <>
      <line
        x1={xScale(first(elements))}
        x2={xScale(last(elements))}
        y1={y}
        y2={y}
        style={{ strokeWidth: '3px', stroke: 'black' }}
      />
      {elements.map((d) => (
        <circle
          cx={xScale(d)}
          cy={y}
          r={4}
          style={{
            fill: 'black',
            strokeWidth: 3,
          }}
          onMouseUp={() => setSliderSelection(d)}
          key={d}
        />
      ))}
      <circle
        cx={
          mousePosition.x !== null
            ? clamp(mousePosition.x, xScale(first(elements)), xScale(last(elements)))
            : xScale(sliderSelection)
        }
        cy={y}
        r={12}
        style={{ fill: 'white', strokeWidth: 3, stroke: 'black' }}
        onMouseDown={() => {
          window.addEventListener('mouseup', stopMovingSlider)
          onSliderReleaseRef.current = stopMovingSlider

          window.addEventListener('mousemove', followMouse)
          onMouseMoveRef.current = followMouse
        }}
      />
    </>
  )
}
