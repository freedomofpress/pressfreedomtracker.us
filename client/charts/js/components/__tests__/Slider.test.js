import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import Slider from '../Slider'

test('renders CheckBoxesYear with mocked data', () => {
	const renderer = new ShallowRenderer();
	expect(renderer.render(
		<Slider
			elements={["Nov", "Dec", "Jan", "Feb", "Mar", "Apr"]}
			xScale={(x) => x}
			y={400}
			sliderSelection={'Nov'}
			idContainer={'barchart-svg'}
		/>
	)).toMatchSnapshot();
});
