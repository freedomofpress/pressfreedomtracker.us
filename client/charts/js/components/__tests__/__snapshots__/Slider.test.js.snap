// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders CheckBoxesYear with mocked data 1`] = `
<React.Fragment>
  <line
    style={
      {
        "stroke": "black",
        "strokeWidth": "3px",
      }
    }
    x1="Nov"
    x2="Apr"
    y1={400}
    y2={400}
  />
  <circle
    cx="Nov"
    cy={400}
    onMouseUp={[Function]}
    r={4}
    style={
      {
        "fill": "black",
        "strokeWidth": 3,
      }
    }
  />
  <circle
    cx="Dec"
    cy={400}
    onMouseUp={[Function]}
    r={4}
    style={
      {
        "fill": "black",
        "strokeWidth": 3,
      }
    }
  />
  <circle
    cx="Jan"
    cy={400}
    onMouseUp={[Function]}
    r={4}
    style={
      {
        "fill": "black",
        "strokeWidth": 3,
      }
    }
  />
  <circle
    cx="Feb"
    cy={400}
    onMouseUp={[Function]}
    r={4}
    style={
      {
        "fill": "black",
        "strokeWidth": 3,
      }
    }
  />
  <circle
    cx="Mar"
    cy={400}
    onMouseUp={[Function]}
    r={4}
    style={
      {
        "fill": "black",
        "strokeWidth": 3,
      }
    }
  />
  <circle
    cx="Apr"
    cy={400}
    onMouseUp={[Function]}
    r={4}
    style={
      {
        "fill": "black",
        "strokeWidth": 3,
      }
    }
  />
  <circle
    cx="Nov"
    cy={400}
    onMouseDown={[Function]}
    r={12}
    style={
      {
        "fill": "white",
        "stroke": "black",
        "strokeWidth": 3,
      }
    }
  />
</React.Fragment>
`;
